### ISSUE TITLE
Could you please provide the training code?

### ISSUE THREAD

**Issue by ZrrSkywalker** [WJ-Chang-42/ASTransformer#2]
Thanks for your excellent and inspirational work!
Could you please release the code for training by Attention Supervision? so that I can further follow and cite your paper~
Many Thanks.

**Comment by Saurav-31** [comment]
@WJ-Chang-42 Wonderful work! Also looking forward to the training code.

**Comment by WJ-Chang-42** [comment]
Thanks for your following!  The training codes in separate files are hard to combine due to the 2-stage training strategy. I might release an easy-to-use version in the future, but now I have been busy with other work. 
However, I could provide the code of loss function. Hope it can help you.
**Attention Loss**
```
def calculate_attention_loss(volume, gt):
    cos = torch.nn.CosineSimilarity(dim=3, eps=1e-08)
    batch = volume.shape[0]
    dim = volume.shape[2]
    losses3 = torch.sum((gt - volume).abs())/(batch*dim*dim*1)
    part1 = torch.sum(torch.abs(1 -cos(volume,gt)))
    part2 = torch.sum(torch.abs(1 - cos(volume.transpose(2,3),gt.transpose(2,3))))
    losses3 = losses3 + (part1+part2)/(batch*dim*1)
    return losses3
```
**Ground Truth Attention Volume**
```
def get_depth_volume(depth, level, normalize = False):
    batch,H,W = depth.shape
    if normalize:
        temp = torch.flatten(depth, 1)
        max_depth = torch.max(temp,1)[0].unsqueeze(1)
        temp = temp/max_depth
    else:
        temp = torch.flatten(depth, 1)
    attention = torch.zeros((batch,1,H*W,H*W)).cuda()
    volume = torch.abs(temp.unsqueeze(2) - temp.unsqueeze(1))
    attention[:,0,:,:] = torch.softmax((level)*(-volume),2)
    return attention
```

**Comment by Saurav-31** [comment]
Thanks for sharing the loss functions. It will definitely help.

**Comment by koutilya-pnvr** [comment]
The initialization of attention in the get_depth_volume function is computationally intensive and no CPU seems to be able to handle such a memory. Can anyone explain the memory configuration of such a setup?

attention tensor is of size: 1 x 1 x 247808 x 247808

**Comment by WJ-Chang-42** [comment]
> The initialization of attention in the get_depth_volume function is computationally intensive and no CPU seems to be able to handle such a memory. Can anyone explain the memory configuration of such a setup?
> 
> attention tensor is of size: 1 x 1 x 247808 x 247808

Thanks for your following! 
The input of “get_depth_volume” is supposed to be down-sampled 16 times from ground truth to fit the size of feature maps in transformer layers. You could use the code as follows.
`label_depth = torch.nn.functional.interpolate(dense_depths, size=[dense_depths.shape[2]//16,dense_depths.shape[3]//16]).squeeze(1) `

**Comment by koutilya-pnvr** [comment]
Thanks a lot for your reply. This solves the greatest bottleneck for my own implementation of the training code. Also, I would appreciate if you can share the training code for the sake of consistency and reproducibility:)

**Comment by koutilya-pnvr** [comment]
Any idea where can we find the imagenet pretrained weights to initialize the transformer encoder in this work with?

**Comment by WJ-Chang-42** [comment]
> Any idea where can we find the imagenet pretrained weights to initialize the transformer encoder in this work with

 For NYU2, we directly use the pre-trained model from https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-vitjx/jx_vit_base_p16_384-83fb41ba.pth. The weights for KITTI were pre-trained by us and could be found in https://drive.google.com/file/d/109CPEDxPxR_1hrwhjd0rOrUlmPtUnRtx/view?usp=sharing

**Comment by ruili3** [comment]
> > Any idea where can we find the imagenet pretrained weights to initialize the transformer encoder in this work with
> 
> For NYU2, we directly use the pre-trained model from https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-vitjx/jx_vit_base_p16_384-83fb41ba.pth. The weights for KITTI were pre-trained by us and could be found in https://drive.google.com/file/d/109CPEDxPxR_1hrwhjd0rOrUlmPtUnRtx/view?usp=sharing

Hi, thank you for sharing the pre-trained model. Can you share the code to load the pre-trained model? Thank you a lot!

**Comment by WJ-Chang-42** [comment]
Thanks for your attention! You could load the pre-train model following:
++++++++++++++++++++++
img_s = (416,544)
model_dict = torch.load('jx_vit_base_p16_384-83fb41ba.pth', map_location='cpu')
head = model_dict['pos_embed'][:,:1]
pos = model_dict['pos_embed'][0,1:]
pos = pos.view(384//16,384//16,-1)
zoom = (26 / 24, 34 / 24, 1)
posemb_grid = torch.as_tensor(scipy.ndimage.zoom(pos, zoom, order=1)).view(1,26*34,-1)
model_dict['pos_embed'] = torch.cat([head, posemb_grid], 1)
self.encoder.model.load_state_dict(model_dict)

Best wishes,
Wenjie Chang
从 Windows 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>发送

发件人: Rui ***@***.***>
发送时间: 2022年3月8日 21:29
收件人: ***@***.***>
抄送: ***@***.***>; ***@***.***>
主题: Re: [WJ-Chang-42/ASTransformer] Could you please provide the training code? (Issue #2)


Any idea where can we find the imagenet pretrained weights to initialize the transformer encoder in this work with

For NYU2, we directly use the pre-trained model from https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-vitjx/jx_vit_base_p16_384-83fb41ba.pth<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frwightman%2Fpytorch-image-models%2Freleases%2Fdownload%2Fv0.1-vitjx%2Fjx_vit_base_p16_384-83fb41ba.pth&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=aoa%2FTm%2FPybwQe5gFecaPb%2FCUh8t3yiKGoq8n6PFlxbw%3D&reserved=0>. The weights for KITTI were pre-trained by us and could be found in https://drive.google.com/file/d/109CPEDxPxR_1hrwhjd0rOrUlmPtUnRtx/view?usp=sharing<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdrive.google.com%2Ffile%2Fd%2F109CPEDxPxR_1hrwhjd0rOrUlmPtUnRtx%2Fview%3Fusp%3Dsharing&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=bq%2FmhFpJkcTLu2heuw%2FLZjQxMEZgkXLFg%2FtxjuF2LE8%3D&reserved=0>

Hi, thank you for sharing the pre-trained model. Can you share the code to load the pre-trained model? Thank you a lot!

―
Reply to this email directly, view it on GitHub<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FWJ-Chang-42%2FASTransformer%2Fissues%2F2%23issuecomment-1061777818&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=AxkH29nfEW8Jwm3sSD4FxhjANUwn3cRY%2FMJgc7ly37Q%3D&reserved=0>, or unsubscribe<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJQRO2AICDJ2PWWKPG5LUZLU65I2PANCNFSM5KBLXVCA&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=3F4jbrc0AKAYvrjPKIMobyGr67zkf%2FVTfCRGHztas3Y%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=U16K5Ir2Z%2BuQ1TBH8vp0eR%2BEl4UJzgtuZAZVA4f3uhA%3D&reserved=0> or Android<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7C%7C3ce00afd579f495f082d08da0107a226%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637823429555045840%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=oS5TDsVAGVVFH0vNk8JQVKfUvEoK8FSw8S5oZ4NSvdg%3D&reserved=0>.

**Comment by lcqf** [comment]
Have you realized the training code  Can you share it  Thank you.
